CQW #15 pagination

MagdaMalinowska opened the pull request:
The goal of this story was to do a pagination for blog posts page. It was supposed to be SEO friendly infinite scroll

### Main AC:
- [x] infinite scroll
- [x] when visiting blog post back button should take us to the same page we came from

### TODO in WIP PR
- [x] scroll up- page 1 is not loaded
- [x] f ('.page:in-viewport:first') or :last depending on scroll direction
- [x] code refactor (many duplications)
- [x] fix codeship
- [x] test on mobile
- [x] back to first page

### TODO- PR feedback
- [x] fix lint issues (too long lines)
- [x] check loading up 3-2-1 it shouldn't jump to 1 page 

## TODO AFTER PR
- [x] change back number of articles per page (30)
- [x] remove visible page numbers
- [x] set less strict rules for loading next/prev page (it should not be vsisble for user. now it is only for testing)

### Video:
[video](https://drive.google.com/drive/u/1/folders/0B7XnlcVxJi8wVmwzSjJLa0JXVTQ)

### Review
- [x] @AdamMankowski 
- [ ] @irene50288 
- [ ] @pjar 
- [x] @kruszczynski 
- [ ] @rpietraszko

MagdaMalinowska commented on source/blog.html.slim:
```diff
@@ -4,7 +4,9 @@ pageable: true
-.m-grid-collection
+.m-grid-collection#posts
+  .page(data-page-no= page_number)
+  = page_number
```
This line will be removed after testing

kruszczynski commented on config.rb:
```diff
@@ -7,7 +7,7 @@
   blog.layout = 'layouts/blog.html.slim'
   blog.paginate = true
-  blog.per_page = 30
+  blog.per_page = 3
```
Wow!

kruszczynski commented:
![fantastisch](https://cloud.githubusercontent.com/assets/2325802/16684609/5083f274-4506-11e6-9727-384d27ee0a86.jpg)

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+$ ->
+
+  win = $(window)
```
It's nice to prefix jQuery objects with `$`

MagdaMalinowska replied:
ok, thanks- will do that

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+
+  win = $(window)
+  url = window.location.href
+  url_arr = url.split('/')
```
In JS it's common to use camel case instead of underscore case.

pjar replied:
Only part after `/` in URL is what you need here.

MagdaMalinowska replied:
Good point! will change it

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+  url_arr = url.split('/')
+
+  if 'blog' in url_arr
+    pageNumber = parseInt(url_arr[url_arr.length-1]) || 1
```
To get last element of an array it's easier to use `#pop` function https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/pop

MagdaMalinowska replied:
Much better! Thanks:)

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+        adjustUrl()
+        if shouldLoadPrev(scroll_pos)
+          is_loading = 1
+          loadPrevious(pageNumber)
```
You can probably put this if statement into a function - lot's of duplication here

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+        adjustUrl()
+        if shouldLoadNext(scroll_pos)
+          is_loading = 1
+          loadFollowing(pageNumber)
```
Maybe `loadNext` makes more sense?

MagdaMalinowska replied:
+1

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+
+    shouldLoadNext = (scroll_pos, maxHeight) ->
+      bottomMargin = $('footer').height() + $('.m-grid-item').height()
+      maxHeight = $(document).height() - win.height() - bottomMargin
```
This calculation could be a function.

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+        addContentAfter(data)
+
+    addContentAfter = (data) ->
+      ($('.no-content').first()).before(data)
```
Maybe `append` to blog posts container would be easier?

MagdaMalinowska replied:
nope :) it's tricky here because we have this hack to work properly with flex box- we have additional empty .no-content items in the end so next articles must be added before them

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+
+    addContentCallback = ->
+      truncate()
+      history.pushState({}, "page #{pageNumber}", page_url())
```
Isn't that an `adjustUrl` job? :)

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,100 @@
+      if pageNumber > 1
+        '/blog/page/'+ pageNumber
+      else
+        '/blog/'
```
`return '/blog/' unless pageNumber > 1`

MagdaMalinowska replied:
+1

pjar commented on source/layouts/posts.html.slim:
```diff
@@ -0,0 +1,25 @@
+    = stylesheet_link_tag 'normalize', 'application'
+    / It is only to recognize touch events:
+    = javascript_include_tag 'lib/modernizr-custom'
+    = javascript_include_tag 'application'
```
Do we need JS to be loaded before content?

MagdaMalinowska replied:
yy.. this one I don't get:) What do you suggest?

pjar replied:
I've been learned that JS should be loaded after content, to make it visible quicker.

MagdaMalinowska replied:
that's probably sth to refactor globally- we have it like this in every layout:/

pjar commented:
Knowing now what `source/javascripts/scroll.js.coffee` should do and how, it's probably good to refactor it in next PR.

MagdaMalinowska commented:
@pjar addressed all issues that you mentioned- fixed/changed/or write reply to comment. Thanks a lot for feedback! 
@AdamMankowski problem with scroll already fixed

pjar commented on source/javascripts/scroll.js.coffee:
```diff
@@ -0,0 +1,104 @@
+
+  if 'blog' in urlArr
+    pageNumber = parseInt(urlArr.pop()) || 1
+    $numPages = $('#total_pages').text()
```
It's not a jquery object :) 

To make sure that thing is a number, you could add plus - `+ $('#total_pages').text()`

MagdaMalinowska replied:
yeah, you're right. With plus i didn't know that- thanks:)

MagdaMalinowska commented:
@pjar merging but let me know if there is sth else that we should address here in your opinion- we will create a tasks when you specify details of AC :)
